model_code.py

```diff
@@ -92,4 +92,181 @@ def Gibbs_regression(X, y, B):
+    p = np.exp(alpha0 + alpha1*age_missing_higher_yes) / (1 + np.exp(alpha0 + alpha1*age_missing_higher_yes))
+    return np.sum((1 - higher_yes) * np.log(1 - p) + higher_yes * np.log(p)) - (alpha1 - loc)**2 / (2 * scale)
+
+
+def dist_log_gamma0(gamma0, gamma1, absences, age_missing_absences, loc, scale):
+    mu = np.exp(gamma0 + gamma1*age_missing_absences)
+    return np.sum(absences * (gamma0 + gamma1*age_missing_absences) - mu) - (gamma0 - loc)**2 / (2 * scale)
+    # np.prod(poisson.pmf(absences, mu)) * norm.pdf(gamma0, loc=0, scale=100)
+
+
+def dist_log_gamma1(gamma1, gamma0, absences, age_missing_absences, loc, scale):
+    mu = np.exp(gamma0 + gamma1*age_missing_absences)
+    return np.sum(absences * (gamma0 + gamma1*age_missing_absences) - mu) - (gamma1 - loc)**2 / (2 * scale)
+    # np.prod(poisson.pmf(absences, mu)) * norm.pdf(gamma1, loc=0, scale=100)
+
+
+def Gibbs_MH(X, y, B, n, higher_yes_col, absences_col, age_col, tau, loc=0, scale=10):
+    """
+    need to add seed -- think it will be worth it
+
+    """
+    ################# Initializations ################
+    # regression parameters
+    betas = np.zeros((X.shape[1], 2 * B))
+    sigmas2 = np.zeros(2 * B)
+
+    # parameters for missing covariates
+    higher_yes_missing_idx = np.where(np.isnan(X[:, higher_yes_col]))[0]
+    higher_yes_sim = np.zeros((len(higher_yes_missing_idx), 2 * B))
+    alphas0, alphas1 = np.zeros(2 * B), np.zeros(2 * B)
+
+    absences_missing_idx = np.where(np.isnan(X[:, absences_col]))[0]
+    absences_sim = np.zeros((len(absences_missing_idx), 2 * B))
+    gammas0, gammas1 = np.zeros(2 * B), np.zeros(2 * B)
+
+    # getting ages for the missing values
+    age_missing_higher_yes = X[higher_yes_missing_idx, age_col]
+    age_missing_absences = X[absences_missing_idx, age_col]
+
+    # initialize regression and missing covariates parameters
+    higher_yes = round(np.nanmean(X[:, higher_yes_col]))
+    higher_yes_sim[:, 0] = higher_yes
+    X[higher_yes_missing_idx, higher_yes_col] = higher_yes_sim[:, 0]
+    alpha0, alpha1 = 0.5, 0.5
+    alphas0[0], alphas1[0] = alpha0, alpha1
+
+    absences = round(np.nanmean(X[:, absences_col]))
+    absences_sim[:, 0] = absences
+    X[absences_missing_idx, absences_col] = absences_sim[:, 0]
+    gamma0, gamma1 = 0.05, 0.05
+    gammas0[0], gammas1[0] = gamma0, gamma1
+
+    reg = sm.OLS(y, X)
+    res = reg.fit()
+    beta_hat = res.params
+    # same as:
+    # np.linalg.inv(X.T.dot(X)).dot(X.T).dot(y)
+    sigma2 = res.mse_total
+    vbeta = np.linalg.inv(X.T.dot(X))
+    betas[:, 0] = beta_hat.copy()
+    sigmas2[0] = sigma2
+
+    accepts_alpha0, accepts_alpha1, accepts_gamma0, accepts_gamma1 = 0, 0, 0, 0
+    ##################################################
+
+    ############ Sampling Gibbs + MH #################
+    proposal_alpha0 = norm(loc=0, scale=tau)
+    proposal_alpha1 = norm(loc=0, scale=tau)
+    proposal_gamma0 = norm(loc=0, scale=tau)
+    proposal_gamma1 = norm(loc=0, scale=tau)
```

Comment:
I'm wondering if these should actually be centered at 0. Would it make more sense to center them at the initial values, or some other estimate of the true value of these parameters?

Reply:
Yeah so far I sort of randomly chose the initial values but I think we should choose better start values? To make mu (poisson) and p (bernoulli) reasonable and not super large as our absences count and p don't seem to be extremely large when looking at the observed absences count and higher_yes. We can center them at these better start values or maybe center them at the previous sampled value? Would proposal_alpha0 = norm(loc = alphas0[i-1], tau) make sense?

Reply:
> I think we should choose better start values?

What are your thoughts on doing a quick logistic regression? Where the target variable is missingness. The coefficients of the logistic regression can be our start values and our centers.

Reply:
I'm more than happy to handle this in a later PR though

Reply:
Pushed the notebook with the logistic. I can let the starting values be those and push again. We still need to think about the other covariate -- assuming normality on G2 does throw nonsensical values. We can go with G2 but maybe one of you wants to work parallel to me and figure out a way of getting positive values.